app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt

```diff
@@ -0,0 +1,154 @@
+package com.mapbox.maps.testapp.examples.markersandcallouts.infowindow
+
+import android.view.View
+import com.mapbox.geojson.Point
+import com.mapbox.maps.MapView
+import com.mapbox.maps.ScreenCoordinate
+import com.mapbox.maps.plugin.annotation.annotations
+import com.mapbox.maps.plugin.annotation.generated.OnPointAnnotationClickListener
+import com.mapbox.maps.plugin.annotation.generated.PointAnnotation
+import com.mapbox.maps.plugin.annotation.generated.PointAnnotationManager
+import com.mapbox.maps.plugin.annotation.generated.createPointAnnotationManager
+import com.mapbox.maps.plugin.gestures.OnMapClickListener
+import com.mapbox.maps.plugin.gestures.addOnMapClickListener
+import com.mapbox.maps.plugin.gestures.removeOnMapClickListener
+import com.mapbox.maps.viewannotation.OnViewAnnotationPositionUpdatedListener
+import com.mapbox.maps.viewannotation.viewAnnotationOptions
+import java.util.concurrent.CopyOnWriteArrayList
+import kotlin.math.abs
+
+/**
+ * Manager class helping to control the [Marker]s.
+ */
+class MarkerManager(
+  private val mapView: MapView
+) : OnPointAnnotationClickListener, OnMapClickListener {
+
+  private val pointAnnotationManager: PointAnnotationManager = mapView.annotations.createPointAnnotationManager()
+  // using copy on write just in case as potentially remove may be called while we're iterating in on click listener
+  private val markerList = CopyOnWriteArrayList<Marker>()
+
+  init {
+    pointAnnotationManager.addClickListener(this)
+    // by adding regular map click listener we implement deselecting all info windows on map click
+    // in legacy code it was controlled by flag in API
+    mapView.getMapboxMap().addOnMapClickListener(this)
+  }
+
+  override fun onAnnotationClick(annotation: PointAnnotation): Boolean {
+    markerList.forEach { marker ->
+      if (marker.pointAnnotation == annotation) {
+        selectMarker(marker, deselectIfSelected = true)
+      }
+    }
+    return true
+  }
+
+  override fun onMapClick(point: Point): Boolean {
+    deselectMarkers()
+    return true
+  }
+
+  fun addMarker(marker: Marker): Marker {
+    marker.prepareAnnotationMarker(pointAnnotationManager)
+    marker.prepareViewAnnotation(mapView)
+    markerList.add(marker)
+    // do not show info window by default
+    deselectMarker(marker)
+    return marker
+  }
+
+  fun removeMarker(marker: Marker) {
+    if (!marker.prepared) {
+      return
+    }
+    markerList.remove(marker)
+    mapView.viewAnnotationManager.removeViewAnnotation(marker.viewAnnotation)
+    pointAnnotationManager.delete(marker.pointAnnotation)
+  }
+
+  fun selectMarker(marker: Marker, deselectIfSelected: Boolean = false) {
+    if (mapView.viewAnnotationManager.getViewAnnotationOptionsByView(marker.viewAnnotation)?.selected == true) {
```

Comment:
nit:

seems to be a good extension
`fun MapView.isMarkerSelected(marker) = viewAnnotationManager.getViewAnnotationOptionsByView(marker.viewAnnotation)?.selected == true`

Reply:
👍